crates/control/src/services/connectors.rs

```diff
@@ -1,17 +1,109 @@
 use std::path::Path;
 
+use anyhow::anyhow;
+use models::{CompositeKey, JsonPointer, Object, RelativeUrl, Schema};
 use tokio::process::Command;
 
 use crate::config::settings;
+use crate::controllers::json_api::RawJson;
+use crate::models::connector_images::ConnectorImage;
+use crate::models::connectors::{Connector, ConnectorOperation};
+use crate::services::subprocess::Subprocess;
 
-pub fn spec(image: &str) -> Command {
+#[derive(Debug, thiserror::Error)]
+pub enum ConnectorError {
+    #[error("could not execute the connector")]
+    ConnectorFailed(anyhow::Error),
+    #[error("failed to parse configuration")]
+    MalformedConfig(serde_json::Error),
+    #[error("could not parse the connector output")]
+    MalformedOutput(serde_json::Error),
+    #[error("connector does not support operation: `{0}`")]
+    UnsupportedOperation(ConnectorOperation),
+}
+
+pub async fn spec(image: &ConnectorImage) -> Result<RawJson, ConnectorError> {
+    // TODO: Swap `image.pinned_version()` out with `image.full_name()`?
+    let image_output = spec_cmd(&image.pinned_version())
+        .execute()
+        .await
+        .map_err(|e| ConnectorError::ConnectorFailed(anyhow!(e)))?;
+    serde_json::from_str(&image_output).map_err(ConnectorError::MalformedOutput)
+}
+
+fn spec_cmd(image: &str) -> Command {
     let mut cmd = Command::new("flowctl");
     cmd.arg("api").arg("spec").arg("--image").arg(image);
 
     cmd
 }
 
-pub fn discovery(image: &str, config_path: &Path) -> Command {
+#[derive(Debug, Deserialize)]
+pub struct DiscoverResponse {
+    pub bindings: Vec<DiscoveredBinding>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DiscoveredBinding {
+    /// A recommended display name for this discovered binding.
+    pub recommended_name: String,
+    /// JSON-encoded object which specifies the endpoint resource to be captured.
+    #[serde(rename = "resourceSpec")]
+    pub resource_spec_json: Object,
+    /// JSON schema of documents produced by this binding.
+    #[serde(rename = "documentSchema")]
+    pub document_schema_json: Object,
+    /// Composite key of documents (if known), as JSON-Pointers.
+    pub key_ptrs: Vec<String>,
+}
+
+impl DiscoveredBinding {
+    pub fn key(&self) -> models::CompositeKey {
+        CompositeKey::new(
+            self.key_ptrs
+                .iter()
+                .map(JsonPointer::new)
+                .collect::<Vec<JsonPointer>>(),
+        )
+    }
+
+    pub fn schema_url(&self) -> Schema {
+        Schema::Url(RelativeUrl::new(self.schema_name()))
+    }
+
+    pub fn schema_name(&self) -> String {
+        format!("{}.schema.json", self.recommended_name)
+    }
+}
+
+pub async fn discover(
```

Comment:
For (near) future PR: we should be running actual discovery in a service loop. The controller should create the _intention_ to run a discovery rather than doing so itself.